# Games - categories

In [1]:
import pandas as pd
import json
import re
import numpy as np

importo games_dataset di twitch:

In [2]:
games_tw = pd.read_csv('https://raw.githubusercontent.com/gianscuri/Twitch_Community_Graph/main/DataProcessing/Games_dataset.csv', sep = ',')
len(games_tw)

1144

Creo dizionario : \
    {'ID_game' : { \
        'twitch_name': ' ', \
         'steam_name' : ' ', \
         'match_score': ' '\
          'twitch_clean': ' ', \
         'steam_clean' : ' ', \
         }

In [3]:
match_dict = {}

for i in range(len(games_tw)):
    match_dict[i] = {}
    match_dict[i]['twitch_name'] = games_tw.game_name[i]
    match_dict[i]['steam_name'] = ''
    match_dict[i]['match_score'] = ''
    match_dict[i]['twitch_clean'] = re.sub(r"[^a-zA-Z0-9 ']",'',match_dict[i]['twitch_name']).lower().strip()
    match_dict[i]['steam_clean'] = ''
match_dict

{0: {'twitch_name': 'Monster Hunter Generations Ultimate',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'monster hunter generations ultimate',
  'steam_clean': ''},
 1: {'twitch_name': 'Alice: Madness Returns',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'alice madness returns',
  'steam_clean': ''},
 2: {'twitch_name': 'Hold Fast',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'hold fast',
  'steam_clean': ''},
 3: {'twitch_name': 'Pokémon Trading Card Game Online',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'pokmon trading card game online',
  'steam_clean': ''},
 4: {'twitch_name': 'Witch It',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'witch it',
  'steam_clean': ''},
 5: {'twitch_name': "Dante's Inferno",
  'steam_name': '',
  'match_score': '',
  'twitch_clean': "dante's inferno",
  'steam_clean': ''},
 6: {'twitch_name': 'The Stanley Parable: Ultra Deluxe',
  'steam_name': '',
  'match_score': '',
  'twitc

In [4]:
with open(r'C:\Users\gross\OneDrive\Desktop\games.json', encoding="utf8") as f:
    games = json.load(f)
#games

In [5]:
games['906850']

{'name': 'Again?',
 'release_date': {'coming_soon': False, 'date': 'Aug 8, 2018'},
 'required_age': 0,
 'is_free': False,
 'price': '$0.49',
 'detailed_description': "Again? is easy game for relax with quiet music. You don't need to worry about achievements, about competition with other players.<br />\r\nThe principle of the game: before you square field 4x4, it randomly appear cubes with the number &quot;2&quot;,by clicking on the arrows to move the cubes together trying to connect the same numbers. Your mission is to get a higher number. But beware, if the field is full, you have to start AGAIN!?<br />\r\nFeatures:<br />\r\n- minimum disk space and easy installation;<br />\r\n- the easiest gameplay;<br />\r\n- wonderful music;<br />\r\n- ideal, who want to kill 5-10 minutes, without dipping into the game.<br />\r\nJust start and play!",
 'about_the_game': "Again? is easy game for relax with quiet music. You don't need to worry about achievements, about competition with other players.

lista giochi di steam:

In [6]:
steam_games = []
for keys in games:
    x = games[keys]['name']
    steam_games.append(x)
len(steam_games)

54946

In [7]:
len(set(steam_games))

54593

Ci sono 353 nomi duplicati nel json steam_games!!!
per ora rimuovo. Dovrei vedere se i dettagli associati al gioco cambiano

In [8]:
steam_games = list(set(steam_games))
#emoticons + chinese 

Ricavo lista elementi comuni = con match perfetto.

In [9]:
len(list(set(games_tw.game_name) & set(steam_games)))

332

Nel match_dict aggiorno con steam_name e match_score = 100 (come percentuale).

In [10]:
for i in range(len(match_dict)):
    if match_dict[i]['twitch_name'] in steam_games:
        match_dict[i]['steam_name'] = match_dict[i]['twitch_name']
        match_dict[i]['match_score'] = '100'
    else:
        pass
match_dict

{0: {'twitch_name': 'Monster Hunter Generations Ultimate',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'monster hunter generations ultimate',
  'steam_clean': ''},
 1: {'twitch_name': 'Alice: Madness Returns',
  'steam_name': 'Alice: Madness Returns',
  'match_score': '100',
  'twitch_clean': 'alice madness returns',
  'steam_clean': ''},
 2: {'twitch_name': 'Hold Fast',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'hold fast',
  'steam_clean': ''},
 3: {'twitch_name': 'Pokémon Trading Card Game Online',
  'steam_name': '',
  'match_score': '',
  'twitch_clean': 'pokmon trading card game online',
  'steam_clean': ''},
 4: {'twitch_name': 'Witch It',
  'steam_name': 'Witch It',
  'match_score': '100',
  'twitch_clean': 'witch it',
  'steam_clean': ''},
 5: {'twitch_name': "Dante's Inferno",
  'steam_name': '',
  'match_score': '',
  'twitch_clean': "dante's inferno",
  'steam_clean': ''},
 6: {'twitch_name': 'The Stanley Parable: Ultra Deluxe',
  'steam_name

In [11]:
match_df = pd.DataFrame.from_dict(match_dict,orient='index')
match_df

twitch_name                         steam_name  \
0     Monster Hunter Generations Ultimate                                      
1                  Alice: Madness Returns             Alice: Madness Returns   
2                               Hold Fast                                      
3        Pokémon Trading Card Game Online                                      
4                                Witch It                           Witch It   
...                                   ...                                ...   
1139    Rise of Nations: Extended Edition  Rise of Nations: Extended Edition   
1140        Paper Mario: The Origami King                                      
1141     The Inheritance of Crimson Manor                                      
1142                               Occult                                      
1143                      Sign of Silence                    Sign of Silence   

     match_score                         twitch_clean steam_clean  
0                 monster hunter generations ultimate              
1            100                alice madness returns              
2                                           hold fast              
3                     pokmon trading card game online              
4            100                             witch it              
...          ...                                  ...         ...  
1139         100     rise of nations extended edition              
1140                     paper mario the origami king              
1141                 the inheritance of crimson manor              
1142                                           occult              
1143         100                      sign of silence              

[1144 rows x 5 columns]

In [12]:
#per risalire alle info nel json dalla stringa 'steam_name'

# for key in games:
#     if games[key]['name'] == 'steam_name':
#         print(games[key])

## STEP 1 - cleaning & rematch

prima di usare fuzzy faccio cleaning su stringhe non ancora matchate: 
- lower
- strip
- remove emoticons/chinese/japanese/korean characters

faccio rematch

In [13]:
# l'unica stringa con carattere cinese o simile è già stata matchata

# for i in range(len(games_tw.game_name)):
#     print(i, re.findall(r'[\u4e00-\u9fff]+',games_tw.game_name[i]))
    
# match_df.loc[346]

In [14]:
#cleaning twitch_name
# j=0
# for elem in match_df.steam_name:
#     if elem == '':
#         match_df.twitch_clean[j] = re.sub(r"[^a-zA-Z0-9 ']",'',match_df.twitch_name[j]).lower().strip()
#     else:
#         match_df.twitch_clean[j] = ''
#     j+=1

cleaning della lista steam_games e trasformo in df (associo nome originale a pulito)

In [103]:
steam_df = pd.DataFrame()
steam_df['steam_name'] = steam_games

steam_clean = []
for string in steam_games:
    new_string = re.sub(r"[^a-zA-Z0-9 ']",'',string).lower().strip()
    steam_clean.append(new_string)
steam_df['steam_clean'] = steam_clean
steam_df = steam_df[1:].reset_index(drop=True)   #rimuovo prima riga vuota

In [104]:
steam_df.head()

steam_name                   steam_clean
0                    Descending                    descending
1             Dungeon Adventure             dungeon adventure
2  Artifact Adventure Gaiden DX  artifact adventure gaiden dx
3  Slide Stories: Bucky's Quest   slide stories bucky's quest
4                    Fatal Core                    fatal core

In [105]:
steam_df[steam_df.steam_clean == '']

steam_name steam_clean
44                                       江湖侠客行            
50                                      小怪物必须死            
117                                         八尾            
133                                       猜拳为尊            
181                                      招募与冒险            
213                                      看不见的爱            
272                                      溜冰的猪少            
374                              그랑 엠파이어 : 아마네            
485                                       什么找茬            
510                                      去死吧蝙蝠            
602                                   マグネクション磁            
768                               《鬼畜战记：金坷垃传说》            
856                                       问道仙途            
902                               恋爱绮谭~不存在的夏天~            
982                             ガリア戦記　雄篇アンナーレス            
1073                                  アルカナリッター            
1330                                      ドトコイ            
1499                                     恐龙大冒险            
1505                                      伊始之战            
1555                                      浪迹三国            
1681                                       记得我            
1789                                     雪国の少女            
1967                                 炎黄战纪之三国烽烟            
2334                                     团长别生气            
2482                                누가 그녀를 죽였나            
2498                                      琴覆天下            
2558                                  机甲雄心：巡天者            
2687                                    光與暗之戀曲            
2771                      Путешествие паладина            
2810                                    克苏鲁异闻录            
3089                                 عالم أريب            
3110                                       超将棋            
3338                                      中国战疫            
3639                                       玄龙山            
3818                                 三国演义之蜀汉传奇            
3891                                 来自女巫森林的法师            
3935                                     尘埃见闻录            
4006                                     游戏：异星            
4048                                武儒绘卷 - 启示录            
4153                               贤者挽歌之马略卡协奏曲            
4198                                   人中怪物多人版            
4284                                      泡沫冬景            
4314                                幻想英雄传 卡片对决            
4411                                     封神纪传奇            
4534                                    黑白米拖更记            
4609                                     只是一座楼            
4615                                       心之檻            
4791                              그레텔과 윈슬로의 별장            
4967                                      全托教师            
5017                                      黎明霞光            
5137                                      剑道仙语            
5230                                    魔法考核冒险            
5321                                       经天纪            
5346                                      像素女友            
5465                               いせかいあどべんちゃー            
5532                Симулятор Сидения на Крыше            
5601                                ВЗЛОМ ЖОПЫ            
5668                                 灰烬行星与填鸭少女            
5858                                    休闲游戏合集            
5874                            怪獣綺譚 朧十夜 空狐万華鏡            
5894                                      飞机女孩            
5900                                 四圣传说之黑暗森林            
5906                                      小小愿景            
5916                                     欢乐消消乐            
5979                                   勇者的十日人生            
6052                                 斗罗大陆-王国印记            
6185                                    猫头鹰和灯塔            
6333                    

rimuovo queste righe cinesi dal dataframe steam_df

In [106]:
steam_df = steam_df[steam_df.steam_clean != '']
len(steam_df)

53986

In [107]:
pd.set_option("display.max_rows", None, "display.max_columns", None)
match_df

twitch_name  \
0                   Monster Hunter Generations Ultimate   
1                                Alice: Madness Returns   
2                                             Hold Fast   
3                      Pokémon Trading Card Game Online   
4                                              Witch It   
5                                       Dante's Inferno   
6                     The Stanley Parable: Ultra Deluxe   
7                                        The Long Drive   
8                                          Cyber Hunter   
9                                           Rogue Lords   
10                      Assassin's Creed IV: Black Flag   
11                                               INSIDE   
12                          DRAGON BALL Z DOKKAN BATTLE   
13                                                 V696   
14                                                Hades   
15                                        Stranded Deep   
16                             Legacy of Kain: Defiance   
17                                               Rounds   
18                                            Drawful 2   
19                             Sensible World of Soccer   
20                                           Atomicrops   
21                                        Slime Rancher   
22                                 Cafe Owner Simulator   
23                              Dodonpachi Resurrection   
24                                                Bingo   
25                                                 Embr   
26                                     Spyro the Dragon   
27                        Final Fantasy VIII Remastered   
28                                Yomawari: Night Alone   
29                                    Hatoful Boyfriend   
30                                    Tomb Raider: Gold   
31                                         Scribble It!   
32                                       Kingdom Hearts   
33                                         Dwarfenstein   
34                                   Gran Turismo Sport   
35                                              Liftoff   
36                                     The Lost Vikings   
37                                           Pupperazzi   
38                                     League Challenge   
39                                           Metal Mind   
40    Dragon Quest XI S: Echoes of an Elusive Age - ...   
41                              Silent Hill: Homecoming   
42                                            Destiny 2   
43                                      Words On Stream   
44                Harry Potter and the Sorcerer's Stone   
45                              Five Nights at Freddy's   
46                               Eyes - The Horror Game   
47                                      Pokémon Scarlet   
48                         Sherlock Holmes: Chapter One   
49                                       Planet Coaster   
50                                           The Forest   
51                                   Salt and Sanctuary   
52                                       Endless Legend   
53                                  Heroes of the Storm   
54                                              Quiz It   
55                                          Inscryption   
56                                 Shin Megami Tensei V   
57                                          Space Punks   
58                          This War of Mine: Final Cut   
59                                            Loop Hero   
60                            The Amulet of AmunRuuUUUN   
61                                              Elsword   
62                               Golf With Your Friends   
63                                               Visage   
64                           Anno 1404: History Edition   
65                                   Salt and Sacrifice   
66                            Quiplash 2 InterLASHional   
67                              

ricavo set elementi comuni con match perfetto tra liste pulite

In [108]:
len(list(set(match_df.twitch_clean) & set(steam_clean)))

412

In [109]:
for i in range(len(match_df)):
    if ((match_df.match_score[i] != '100') and (match_df.twitch_clean[i] in list(set(steam_clean)))):
        match_df.steam_clean[i] = match_df.twitch_clean[i]
        match_df.match_score[i] = '100'
    else:
        pass
match_df

twitch_name  \
0                   Monster Hunter Generations Ultimate   
1                                Alice: Madness Returns   
2                                             Hold Fast   
3                      Pokémon Trading Card Game Online   
4                                              Witch It   
5                                       Dante's Inferno   
6                     The Stanley Parable: Ultra Deluxe   
7                                        The Long Drive   
8                                          Cyber Hunter   
9                                           Rogue Lords   
10                      Assassin's Creed IV: Black Flag   
11                                               INSIDE   
12                          DRAGON BALL Z DOKKAN BATTLE   
13                                                 V696   
14                                                Hades   
15                                        Stranded Deep   
16                             Legacy of Kain: Defiance   
17                                               Rounds   
18                                            Drawful 2   
19                             Sensible World of Soccer   
20                                           Atomicrops   
21                                        Slime Rancher   
22                                 Cafe Owner Simulator   
23                              Dodonpachi Resurrection   
24                                                Bingo   
25                                                 Embr   
26                                     Spyro the Dragon   
27                        Final Fantasy VIII Remastered   
28                                Yomawari: Night Alone   
29                                    Hatoful Boyfriend   
30                                    Tomb Raider: Gold   
31                                         Scribble It!   
32                                       Kingdom Hearts   
33                                         Dwarfenstein   
34                                   Gran Turismo Sport   
35                                              Liftoff   
36                                     The Lost Vikings   
37                                           Pupperazzi   
38                                     League Challenge   
39                                           Metal Mind   
40    Dragon Quest XI S: Echoes of an Elusive Age - ...   
41                              Silent Hill: Homecoming   
42                                            Destiny 2   
43                                      Words On Stream   
44                Harry Potter and the Sorcerer's Stone   
45                              Five Nights at Freddy's   
46                               Eyes - The Horror Game   
47                                      Pokémon Scarlet   
48                         Sherlock Holmes: Chapter One   
49                                       Planet Coaster   
50                                           The Forest   
51                                   Salt and Sanctuary   
52                                       Endless Legend   
53                                  Heroes of the Storm   
54                                              Quiz It   
55                                          Inscryption   
56                                 Shin Megami Tensei V   
57                                          Space Punks   
58                          This War of Mine: Final Cut   
59                                            Loop Hero   
60                            The Amulet of AmunRuuUUUN   
61                                              Elsword   
62                               Golf With Your Friends   
63                                               Visage   
64                           Anno 1404: History Edition   
65                                   Salt and Sacrifice   
66                            Quiplash 2 InterLASHional   
67                              

In [110]:
match_df[match_df.steam_clean != ''].count()

twitch_name     80
steam_name      80
match_score     80
twitch_clean    80
steam_clean     80
dtype: int64

80 nuovi elementi matchati -> inserisco in dataframe con 100%

In [111]:
len(steam_df.steam_clean) - len(set(steam_df.steam_clean))

273

ci sono 273 duplicati in lista steam_clean --> elimino

In [114]:
steam_df = steam_df.drop_duplicates(subset=['steam_clean'], keep='first').reset_index(drop=True)
len(steam_df)

53713

In [116]:
steam_df.head()

steam_name                   steam_clean
0                    Descending                    descending
1             Dungeon Adventure             dungeon adventure
2  Artifact Adventure Gaiden DX  artifact adventure gaiden dx
3  Slide Stories: Bucky's Quest   slide stories bucky's quest
4                    Fatal Core                    fatal core

In [124]:
############# COSA ETERNA!!!!

In [122]:
contatore = 0
for i in range(len(match_df)):
    for j in range(len(steam_df)):
        if match_df.steam_clean[i]  == steam_df.steam_clean[j]:
            match_df.steam_name[i] = steam_df.steam_name[j]   
    contatore += 1
    print(f"sono alla {contatore} richiesta", end ="\r")

In [126]:
#match_df.to_csv('match_df.csv')

###### salvato df match perfetti

In [132]:
match_df[match_df.match_score == '100'].to_csv('match_100.csv')

In [135]:
match_df[match_df.match_score != '100'].to_csv('non_match.csv')

In [142]:
fuzzy_match = pd.read_csv('C:\\Users\\gross\\OneDrive\\Documenti\\GitHub\\Twitch_Community_Graph\\DataProcessing\\non_match.csv',sep=',')
fuzzy_match = fuzzy_match.replace(np.nan, '', regex=True).drop(['Unnamed: 0'],axis=1)
fuzzy_match

twitch_name steam_name match_score  \
0                  Monster Hunter Generations Ultimate                          
1                                            Hold Fast                          
2                     Pokémon Trading Card Game Online                          
3                                      Dante's Inferno                          
4                    The Stanley Parable: Ultra Deluxe                          
5                                       The Long Drive                          
6                      Assassin's Creed IV: Black Flag                          
7                          DRAGON BALL Z DOKKAN BATTLE                          
8                                                Hades                          
9                             Sensible World of Soccer                          
10                                Cafe Owner Simulator                          
11                                               Bingo                          
12                                    Spyro the Dragon                          
13                       Final Fantasy VIII Remastered                          
14                                   Tomb Raider: Gold                          
15                                        Scribble It!                          
16                                      Kingdom Hearts                          
17                                        Dwarfenstein                          
18                                  Gran Turismo Sport                          
19                                             Liftoff                          
20                                    The Lost Vikings                          
21                                    League Challenge                          
22                                          Metal Mind                          
23   Dragon Quest XI S: Echoes of an Elusive Age - ...                          
24                                           Destiny 2                          
25                                     Words On Stream                          
26               Harry Potter and the Sorcerer's Stone                          
27                              Eyes - The Horror Game                          
28                                     Pokémon Scarlet                          
29                                 Heroes of the Storm                          
30                                             Quiz It                          
31                                Shin Megami Tensei V                          
32                                         Space Punks                          
33                         This War of Mine: Final Cut                          
34                           The Amulet of AmunRuuUUUN                          
35                              Golf With Your Friends                          
36                          Anno 1404: History Edition                          
37                                  Salt and Sacrifice                          
38                                             Valheim                          
39                               Batman: Arkham Asylum                          
40                                           Marauders                          
41                                         World War 3                          
42                   Metal Gear Solid V: Ground Zeroes                          
43                     Call of Duty: Black Ops Zombies                          
44                                 Super Mario Odyssey                          
45                                  Escape from Tarkov                          
46                                   Truck Simulator 2                          
47                    Call of Duty: Black Ops Cold War                          
48                                    Fitness & Health                    

### fuzzymatcher

In [ ]:
# fts4 in DDL
# >>> import os
# >>> import sys
# >>> os.path.dirname(sys.executable)

In [ ]:
import fuzzymatcher

In [ ]:
# cosine distance -> non pesa l'ordine
# levenshtein -> numero modifiche per passare da A a B
            # -> d(lev) >= |len(A)-len(B)|

In [ ]:
matched_results = fuzzymatcher.fuzzy_left_join(match_df,
                                               steam_df,
                                              left_on='twitch_clean',
                                              right_on= 'steam_clean')

In [ ]:
#solo 812 non ancora matchati

matched_results[matched_results['twitch_clean'] != ''][['best_match_score','twitch_clean','steam_clean','steam_name_right']].sort_values(by=['best_match_score'], ascending=False)

In [ ]:
matched_results.loc[6836]
#cambia tutte le volte che rirunno

In [ ]:
matched_results.twitch_clean.loc[6836] == matched_results.steam_clean.loc[6836]

match perfetto ma 0.28 best_match_score!!!

In [ ]:
#match_df['steam_clean'] = ''

# j=0
# for elem in matched_results.twitch_clean:
#     if elem == '':
#         pass
#     else:
#         match_df.steam_clean[j] = matched_results.steam_clean
#         match_df.[j]
#     j+=1

capire come convertire best_match_score in percentuale  -> non si può schifo di documentazione

## passo a fuzzywuzzy process

In [ ]:
# pip install fuzzywuzzy
from fuzzywuzzy import process

In [ ]:
def fuzzy_merge(df_1, df_2, key1, key2, threshold=90, limit=2):
    """
    :param df_1: the left table to join
    :param df_2: the right table to join
    :param key1: key column of the left table
    :param key2: key column of the right table
    :param threshold: how close the matches should be to return a match, based on Levenshtein distance
    :param limit: the amount of matches that will get returned, these are sorted high to low
    :return: dataframe with boths keys and matches
    """
    s = df_2[key2].tolist()
    
    m = df_1[key1].apply(lambda x: process.extract(x, s, limit=limit))    
    df_1['matches'] = m
    
    m2 = df_1['matches'].apply(lambda x: ', '.join([i[0] for i in x if i[1] >= threshold]))
    df_1['matches'] = m2
    
    return df_1

In [ ]:
fuzzy_merge(match_df, steam_df, 'twitch_clean', 'steam_clean', threshold=80)